### PR TITLE
[FLINK-27171][runtime][security] Add periodic kerberos delegation token obtain possibility to DelegationTokenManager

### PR DESCRIPTION
**Pull request by gaborgsomogyi**
## What is the purpose of the change

Kerberos delegation token framework is now not able to obtain and periodically re-obtain tokens from token providers. In this PR I've added such functionality.

## Brief change log

* Added `security.kerberos.tokens.renewal-ratio` with default 0.75
* Added `security.kerberos.tokens.retry-wait` with default 1 hour
* Added one-time delegation token obtain possibility in `obtainDelegationTokens`
* Added re-occurring delegation token obtain possibility in `start`
* Minor renames/fixes here and there

## Verifying this change

* Additional unit tests
* Manually on YARN and K8S (please see an example [here](https://gist.github.com/gaborgsomogyi/ac4f71ead8494da2f5c35265bcb1e885))

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? All documentation is intended to be added in [FLINK-25911](https://issues.apache.org/jira/browse/FLINK-25911) when everything works as a whole

### CODE REVIEW

**Comment by gaborgsomogyi** [comment]
cc @mbalassi

**Comment by gaborgsomogyi** [comment]
Here is an example run log from K8S.
[token.log](https://github.com/apache/flink/files/8778806/token.log)

**Comment by mbalassi** [comment]
Thanks, @gaborgsomogyi. I glanced through the log, looks good at first sight. I will be able to give you detailed feedback early next week.

**Comment by gaborgsomogyi** [comment]
After a minor fix here is the TGT based log.
[token_tgt.log](https://github.com/apache/flink/files/8780948/token_tgt.log)

**Comment by gaborgsomogyi** [comment]
@flinkbot run azure

**Comment by gaborgsomogyi** [comment]
@JackWangCS FYI, at top of this this PR you can implement and test FLINK-25909.

**Comment by gaborgsomogyi** [comment]
@flinkbot run azure

File: flink-core/src/main/java/org/apache/flink/configuration/SecurityOptions.java [hunk 1]
```diff
@@ -132,6 +132,22 @@ public class SecurityOptions {
                     .withDescription(
                             "The time period when keytab login happens automatically in order to always have a valid TGT.");
 
+    @Documentation.Section(Documentation.Sections.SECURITY_AUTH_KERBEROS)
+    public static final ConfigOption<Duration> KERBEROS_TOKENS_RETRY_WAIT =
+            key("security.kerberos.tokens.retry-wait")
```
Comment (gyfora) [inline flink-core/src/main/java/org/apache/flink/configuration/SecurityOptions.java:137]: I think a better name would be `security.kerberos.tokens.obtain.retry.backoff` or `security.kerberos.tokens.retry.backoff`

File: flink-core/src/main/java/org/apache/flink/configuration/SecurityOptions.java [hunk 2]
```diff
@@ -132,6 +132,22 @@ public class SecurityOptions {
...
+                    .durationType()
+                    .defaultValue(Duration.ofHours(1))
+                    .withDescription(
+                            "The time period how long to wait before retrying to obtain new delegation tokens after a failure.");
+
+    @Documentation.Section(Documentation.Sections.SECURITY_AUTH_KERBEROS)
+    public static final ConfigOption<Double> KERBEROS_TOKENS_RENEWAL_RATIO =
+            key("security.kerberos.tokens.renewal-ratio")
```
Comment (gyfora) [inline flink-core/src/main/java/org/apache/flink/configuration/SecurityOptions.java:145]: A clearer name would be `security.kerberos.tokens.renewal.time-ratio`

File: flink-runtime/src/main/java/org/apache/flink/runtime/security/token/KerberosDelegationTokenManager.java [hunk 3]
```diff
@@ -127,8 +170,62 @@ boolean isProviderLoaded(String serviceName) {
...
+            LOG.info("Real user has no kerberos credentials so no tokens obtained");
+        }
+    }
+
+    protected long obtainDelegationTokensAndGetNextRenewal(Credentials credentials) {
+        AtomicLong nextRenewal = new AtomicLong(Long.MAX_VALUE);
+
+        delegationTokenProviders
+                .values()
+                .forEach(
```
Comment (gyfora) [inline flink-runtime/src/main/java/org/apache/flink/runtime/security/token/KerberosDelegationTokenManager.java:197]: You could use a stream, flatmap -> renewal time, and min to get rid of the AtomicLong based min logic

File: flink-core/src/main/java/org/apache/flink/configuration/SecurityOptions.java (hunk 1 above)
Comment (gaborgsomogyi) [reply flink-core/src/main/java/org/apache/flink/configuration/SecurityOptions.java:137]: Renamed to `security.kerberos.tokens.retry.backoff`.

File: flink-core/src/main/java/org/apache/flink/configuration/SecurityOptions.java (hunk 2 above)
Comment (gaborgsomogyi) [reply flink-core/src/main/java/org/apache/flink/configuration/SecurityOptions.java:145]: Renamed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/security/token/KerberosDelegationTokenManager.java (hunk 3 above)
Comment (gaborgsomogyi) [reply flink-runtime/src/main/java/org/apache/flink/runtime/security/token/KerberosDelegationTokenManager.java:197]: I've tried this in the initial implementation but it was just not working. No idea what was the reason so re-checking...

Comment (gyfora) [reply flink-runtime/src/main/java/org/apache/flink/runtime/security/token/KerberosDelegationTokenManager.java:197]: If you paste the code here we can probably spot the problem together.

Comment (gaborgsomogyi) [reply flink-runtime/src/main/java/org/apache/flink/runtime/security/token/KerberosDelegationTokenManager.java]: Fixed.

File: flink-core/src/main/java/org/apache/flink/configuration/SecurityOptions.java [hunk 4]
```diff
@@ -132,6 +132,22 @@ public class SecurityOptions {
                     .withDescription(
                             "The time period when keytab login happens automatically in order to always have a valid TGT.");
 
+    @Documentation.Section(Documentation.Sections.SECURITY_AUTH_KERBEROS)
+    public static final ConfigOption<Duration> KERBEROS_TOKENS_RETRY_BACKOFF =
+            key("security.kerberos.tokens.retry.backoff")
```
Comment (gyfora) [inline flink-core/src/main/java/org/apache/flink/configuration/SecurityOptions.java:137]: sorry just to connect the 2 configs maybe even better would be `security.kerberos.tokens.renewal.retry.backoff` so they have a common prefix

Comment (gaborgsomogyi) [reply flink-core/src/main/java/org/apache/flink/configuration/SecurityOptions.java]: Changed.

**Comment by gaborgsomogyi** [comment]
@flinkbot run azure

**Comment by gaborgsomogyi** [comment]
@flinkbot run azure

File: flink-runtime/src/main/java/org/apache/flink/runtime/security/token/KerberosDelegationTokenManager.java [hunk 5]
```diff
@@ -173,27 +274,95 @@ void startTGTRenewal() throws IOException {
...
         if (tgtRenewalFuture != null) {
             tgtRenewalFuture.cancel(true);
             tgtRenewalFuture = null;
         }
     }
 
+    @VisibleForTesting
+    void startTokensUpdate() {
+        try {
+            LOG.info("Starting tokens update task");
```
Comment (gyfora) [inline flink-runtime/src/main/java/org/apache/flink/runtime/security/token/KerberosDelegationTokenManager.java:297]: Should we check `isLoginPossible` before starting this logic?

Comment (gaborgsomogyi) [reply flink-runtime/src/main/java/org/apache/flink/runtime/security/token/KerberosDelegationTokenManager.java:297]: This is an internal function and not intended to use from the outside.
In `start` function there is a check:
```
        if (!kerberosRenewalPossibleProvider.isRenewalPossible()) {
            LOG.info("Renewal is NOT possible, skipping to start renewal task");
            return;
        }
```
What started to bother me though is that `kerberosRenewalPossibleProvider.isRenewalPossible` contains more or less the same code just like `kerberosLoginProvider.isLoginPossible`.

I think debug messages must be added to `kerberosLoginProvider.isLoginPossible` and `kerberosRenewalPossibleProvider` can be deleted as-is. WDYT?

Comment (gaborgsomogyi) [reply flink-runtime/src/main/java/org/apache/flink/runtime/security/token/KerberosDelegationTokenManager.java:294]: I've made the changes, plz share your opinion if you have better idea.

Comment (gyfora) [reply flink-runtime/src/main/java/org/apache/flink/runtime/security/token/KerberosDelegationTokenManager.java:294]: I agree if they contain the same check/logic we should keep only the `kerberosLoginProvider`

Comment (gyfora) [reply flink-runtime/src/main/java/org/apache/flink/runtime/security/token/KerberosDelegationTokenManager.java:294]: looks good

**Comment by gaborgsomogyi** [comment]
I've made a rebase to the latest master because there are several unrelated issues in the jenkins job.

**Comment by gaborgsomogyi** [comment]
@flinkbot run azure

**Comment by gaborgsomogyi** [comment]
FLINK-27792 issue occurs in the jenkins jobs permanently.

**Comment by gaborgsomogyi** [comment]
@flinkbot run azure

**Comment by gaborgsomogyi** [comment]
@flinkbot run azure

**Comment by gaborgsomogyi** [comment]
@flinkbot run azure

**Comment by gaborgsomogyi** [comment]
In the meantime found a real unit test bug which was added. `HadoopReduceFunctionITCase` because of dependeny problem. Namely `HdfsConfiguration` was needed in core because `isProxyUser` was added to a class in `flink-hadoop-fs`. Now only the originally planned `hadoop-common` is needed as expected because I've moved isProxyUser` to `HadoopUserUtils`.

**Comment by JackWangCS** [comment]
> @JackWangCS FYI, at top of this this PR you can implement and test [FLINK-25909](https://issues.apache.org/jira/browse/FLINK-25909).

Thanks, I will start working on the ticket.

**Pull request merged**